src/active_item_cache.cpp

```diff
@@ -7,72 +7,75 @@
 #include "debug.h"
 #include "item.h"
 #include "item_stack.h"
+#include "safe_reference.h"
 
-void active_item_cache::remove( item_stack::iterator it, point location )
+void active_item_cache::remove( const item *it )
 {
-    const auto predicate = [&]( const item_reference & active_item ) {
-        return location == active_item.location && active_item.item_iterator == it;
-    };
-    // The iterator is expected to be in this list, as it was added that way in `add`.
-    // But the processing_speed may have changed, so if it's not in the expected container,
-    // remove it from all containers to ensure no stale iterator remains.
-    auto &expected = active_items[it->processing_speed()];
-    const auto iter = std::find_if( expected.begin(), expected.end(), predicate );
-    if( iter != expected.end() ) {
-        expected.erase( iter );
-    } else {
-        for( auto &e : active_items ) {
-            e.second.remove_if( predicate );
-        }
-    }
-    if( active_item_set.erase( &*it ) == 0 ) {
-        // map erase returns number of elements erased
-        debugmsg( "The item isn't there!" );
-    }
+    active_items[it->processing_speed()].remove_if( [it]( const item_reference & active_item ) {
+        item *const target = active_item.safe_ref.get();
+        return !target || target == it;
+    } );
 }
 
-void active_item_cache::add( item_stack::iterator it, point location )
+void active_item_cache::add( safe_reference<item> it, point location )
 {
-    if( has( it, location ) ) {
+    item *const target = it.get();
+    if( !target ) {
+        debugmsg( "Tried to add nonexistant item to the active item cache at (%d, %d)", location.x,
+                  location.y );
         return;
     }
-    active_items[it->processing_speed()].push_back( item_reference{ location, it, &*it } );
-    active_item_set[ &*it ] = false;
-}
-
-bool active_item_cache::has( item_stack::iterator it, point ) const
-{
-    return active_item_set.find( &*it ) != active_item_set.end();
+    // If the item is alread in the cache for some reason, don't add a second reference
+    std::list<item_reference> &target_list = active_items[target->processing_speed()];
+    if( std::find_if( target_list.begin(),
+    target_list.end(), [target]( const item_reference & active_item_ref ) {
+    return target == active_item_ref.safe_ref.get();
+    } ) != target_list.end() ) {
+        return;
+    }
+    target_list.emplace_back( item_reference{ location, it } );
```

Comment:
```suggestion
    target_list.push_back( item_reference{ location, it } );
```
If you're constructing an object of the container's value_type then there's no need to `emplace_back` (though it's harmless to).

Reply:
done